### PR TITLE
fix(vite): clear a component's styles when it loses its last one

### LINKED ISSUE #457: Removing a component's last style leaves the old CSS applied until a full reload
Split out of a Codex review finding on #455. Pre-existing on `main` — the transition behaves identically there.

## Symptom

A component has styles, then loses them — its last inline `styles` entry is deleted, its array becomes `[]`, or its external stylesheet becomes empty. HMR updates the component, but the old CSS stays applied. Only a full reload clears it.

## Cause

The HMR module spreads the existing `Class.ɵcmp`, so a property it does not emit keeps its previous value. The generator omits `styles` for BOTH null and an empty array:

```
compileForHmrSync(styles = null)        →  no styles property emitted
compileForHmrSync(styles = [])          →  no styles property emitted
compileForHmrSync(styles = ['.a{}'])    →  styles: [ … ]
```

So there is no way for the plugin to express "this component now has no styles" — the distinction is erased on the Rust side.

## Why the obvious fix is not enough

The review that surfaced this recommended preserving a definitive empty state through `compileForHmrSync` and emitting `styles: []`. The first half is a plugin change; the second half is a **Rust generator** change, and without it the plugin cannot express the state no matter what it passes. Both halves are needed.

## Suggested fix

1. Make the HMR generator emit `styles: []` when it is given an empty array, so the spread is overridden, while continuing to omit the property for null (meaning "unknown, leave as-is").
2. Have the plugin pass `[]` rather than null when it knows the component is definitively styleless — #455 already computes that state and currently collapses it to null.

Worth adding endpoint regressions for both transitions, non-empty to empty external and non-empty to empty inline. #455's tests only cover components that were already styleless, so they cannot catch this.

### PR DESCRIPTION
**Pull request by Brooooooklyn**
Closes #457.

A component that lost its last style kept the old CSS applied until a full reload.

## Cause

The generated HMR module opens with `...ClassName.ɵcmp`, so every key it does not emit keeps its previous value. `styles` was omitted for every "no styles" answer, so the stale array survived.

## The issue named one collapse point. There were four.

```
napi/angular-compiler/vite-plugin/index.ts
  :719  readStyles -> null when nothing was read
  :730  (await readStyles(urls)) ?? []          erases the read failure
  :734  merged.length > 0 ? merged : null       <- never sends [] at all

napi/angular-compiler/src/lib.rs
  :555  styles.unwrap_or_default()              None and Some([]) merge here
  :561  if all_styles.is_empty() { None }
  :577  if styles.is_empty() { None }

crates/oxc_angular_compiler/src/hmr/update_module.rs
  :182  if !styles.is_empty()                   second, independent guard
```

Fixing only the generator changes nothing, because `lib.rs:555` never hands it a `Some([])` and the plugin never sends one.

Measured before the fix: `null`, `undefined`, `[]`, `[""]`, `["   "]` and the omitted argument all produced byte-identical output with no `styles:` key.

## The fix

`styles` is now three-valued, end to end:

| value | meaning | module |
|---|---|---|
| non-empty | these are the component's styles | `styles: [ … ]` |
| `[]` | it definitively has none | `styles: []` — clears |
| `null` | unknown | key omitted — the spread keeps what it had |

## `null` was overloaded, and that is the part worth reviewing

In the plugin, `null` meant three different things:

```
(a) the class declares no styles, or an empty array    -> MUST clear
(b) the decorator is unreadable, no fallback           -> MUST keep the old CSS
(c) styleUrls listed but no file could be read         -> was ambiguous
```

Case (b) is why the property cannot simply always be emitted — clearing on an unreadable decorator would wipe live CSS. **So the distinction that matters is whether the answer is KNOWN, not whether it is empty.**

I decided case (c) rather than leaving it open: **a read failure is not evidence of stylelessness.** `readStyles` previously swallowed failures in a `catch {}` and `:730` coalesced the result to `[]`, which made a *missing* stylesheet indistinguishable from an *empty* one. Without the distinction, a truncate window during an editor's atomic write would report the component styleless and wipe its CSS.

```
read succeeded, content empty or whitespace  -> definitively styleless -> clears
read FAILED (missing, permission, throw)     -> unknown                -> left alone
```

The file-level fallback branch keeps its old semantics and can never clear.

## Why `styles: []` actually works

Traced through the pinned Angular v22.0.0 submodule:

```
ɵɵreplaceMetadata (hmr.ts:81) -> recreateLView (hmr.ts:244-312)
  :281  destroyLView(old) -> renderer.destroy()
          -> dom_renderer.ts:610-617  sharedStylesHost.removeStyles(...)
          -> shared_styles_host.ts:184-200  record.usage--; if (<=0) removeElements(...)
  :292  clearRendererCache(oldDef)
  :296  createRenderer(host, newDef) -> this.styles = shim(newDef.styles)   usage++
```

Styles are ref-counted per CSS string. Today the removed and re-added strings are the same, so the count dips and returns and the `<style>` element never leaves. With `styles: []` nothing is re-added, the count reaches zero and `element.remove()` runs. `definition.ts:365` reads `componentDefinition.styles || EMPTY_ARRAY`, and an empty array is truthy, so an emitted `[]` survives intact.

Two caveats found while reading that path, neither blocking: removal is skipped while `allLeavingAnimations.size !== 0` (`dom_renderer.ts:614`), and it is gated on `removeStylesOnCompDestroy`, default true.

## Verification

Binding contract after the fix:

```
null / undefined         -> key omitted
[] / [""] / ["   "]      -> styles: []
[".a{color:red}"]        -> styles: [ … ]
null + template <style>  -> styles: [ … ]     unchanged
```

| | before | after |
|---|---|---|
| `cargo test --lib hmr::` | 16 | 17 |
| `cargo test -p oxc_angular_compiler` | all ok | all ok, 0 failed |
| `pnpm test` | 403 | 408 |
| `pnpm test:e2e` | 36 | 37 |

The five new vitest cases cover both directions — the three clearing transitions, and the two that must NOT clear (an unreadable stylesheet, an unreadable style field).

## One existing assertion changed deliberately

`test/hmr-hot-update.test.ts`, in `serves no styles to a class that declares none, even beside a styled sibling`, asserted `expect(body).not.toContain('styles:')`. That encoded the old behavior. Its real intent — this class is not served its sibling's stylesheet — is still asserted by the marker check; the assertion now pins the explicit empty array instead. Two neighbouring tests passed either way and were tightened the same way.

## The e2e spec is the one that proves the user-visible fix

`e2e/tests/hmr-style-removal.spec.ts` loads a component with an inline style, confirms the dashed border and block host, empties `styles: []`, waits for HMR, and asserts the border reverts to `none` and the host to `inline` — with a DOM sentinel proving the browser did not reload and the websocket wire proving the server never asked for one.

Red-checked: with the generator guard restored and the binary rebuilt, it fails with `Expected: "none" / Received: "dashed"`.

## Note for the reviewer

Two transitions the issue lists still take a full reload rather than the hot path: `styleUrls: ['x'] -> []`, and removing the `styles:` key outright. `stripComponentMetadata` blanks `template:` and `styles:` but not `styleUrls:`, so those edits break the byte-equality check and force a reload. That masking is pre-existing, is not a wrong result, and is left alone here.

---

> [!NOTE]
> **Medium Risk**
> Changes dev-only HMR style resolution; wrong unknown-vs-empty handling could clear live CSS during failed reads, though the new `complete` flag and tests target that edge case.
> 
> **Overview**
> Fixes **#457**: after HMR, a component that went from having styles to none kept old CSS until a full reload, because update modules spread `...ClassName.ɵcmp` and omitted `styles` for every “no styles” case.
> 
> **End-to-end contract:** `styles` is now three-valued — non-empty arrays update CSS, **`[]` means “definitely no styles”** and must emit **`styles: []`** to clear the runtime, and **`null` means unknown** so the key stays omitted and existing styles are preserved.
> 
> The **HMR module generator** emits `styles: []` when given `Some([])` instead of skipping the property. **`compile_for_hmr_sync`** no longer folds definitive empty caller input into `None` after encapsulation. The **Vite plugin** tracks stylesheet read **completeness** (failed/missing reads vs successful empty files) so atomic-write gaps are not treated as “clear styles,” while still allowing partial updates when one of several `styleUrls` fails.
> 
> **API docs** (`index.d.ts`, Rust doc comments) describe the semantics. **Tests** add a Rust unit case, a Playwright e2e for visible CSS removal without reload, and many Vitest scenarios for clear vs must-not-clear paths.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8ae92a155bca199e66e75b3d490af3fda99791ac. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Brooooooklyn** [comment]
The medium finding is confirmed and fixed in 38812d0. The second observation is real but pre-existing, and I am leaving it.

## Confirmed: the fallback branch could clear

Measured at the binding on `c6e339d`:

```
compileForHmrSync(..., [''])    -> styles: []      CLEARS
compileForHmrSync(..., ['  '])  -> styles: []      CLEARS
compileForHmrSync(..., null)    -> key omitted
```

`fallback.contents.length > 0` counted an empty string as content, so a file-level stylesheet that read fine but held nothing was passed on, went definitive at the binding, and came back as `styles: []`. On `088e2ab` the same input produced an omitted key, so this PR introduced it. The branch's own comment already said it never clears — the code had stopped keeping that promise.

The severity is right for the reason you gave: this branch only runs when `classStyles === null`, meaning the class's own decorator was unreadable, and `styleUrls` there is the file-level union across every `@Component` in the file. So the CSS it would erase can belong to a decorator nothing on that path could read.

Fix:

```ts
const fallback = await readStyles(styleUrls)
const usable = fallback.contents.filter((style) => style.trim().length > 0)
styles = usable.length > 0 ? usable : null
```

The whitespace filter now runs BEFORE the null check on this branch, unlike the class-level branch above, where an empty result IS a definitive answer. That asymmetry is the whole point and is now commented.

Two regressions added, both red-checked. Each is one file with two classes: the requested class declares `styles: PS457_FB_STYLES` so the fallback branch runs, and the only file-level stylesheet belongs to the sibling and is empty in one test, whitespace-only in the other. Both assert the served module contains no `styles:` at all. The failure before the fix was `expected '// HMR update for: …' not to contain 'styles:'` with `styles: [],` in the body.

The three clearing cases still clear. Suite: 408 -> 410.

## Declined: the partial-read replacement

> A partial fallback read similarly replaces the full style set with incomplete contents.

True, and unchanged from `main`. The old code on `088e2ab`:

```ts
} catch {
  // Style file not found, continue without this style
}
...
return styleContents.length > 0 ? styleContents : null
```

with a bare `styles = await readStyles(styleUrls)` at the call site. Two stylesheets, one failing, returned the one that read — the subset presented as the whole set, exactly as today.

I am not adding a `fallback.complete` guard here. It would change behaviour this PR did not break, and the underlying problem is that the file-level union is not a per-component answer at all. That is tracked in #456, which removes the fallback rather than narrowing it. Tightening it inside a bug-fix PR would be scope drift.

## On the framing

The recommendation to "keep `styles` null whenever per-class extraction is unknown unless a complete per-class inventory is obtained from the compiler" is where the design is already headed — #456 is exactly that. What this PR can honestly guarantee is narrower: an unknown answer must never become a definitive clear. That invariant now holds on every path, which is what the finding was really about.

File: napi/angular-compiler/vite-plugin/index.ts [hunk 1]
```diff
@@ -727,17 +749,28 @@ export function angular(options: PluginOptions = {}): Plugin[] {
...
+                  // Definitive only when the class's own decorator was read
+                  // (`classStyles !== null`) AND every stylesheet it names was
+                  // read too. An empty result then means the component really
+                  // has no styles and must be cleared — including the case
+                  // where its one stylesheet is now empty or whitespace.
+                  if (external.complete) {
+                    styles = [...classStyles.inline, ...external.contents].filter(
+                      (style) => style.trim().length > 0,
+                    )
+                  }
```
[the remark this answers is not in the document]
Comment (Brooooooklyn) [reply napi/angular-compiler/vite-plugin/index.ts]: Confirmed and fixed in 8ae92a1. You caught a regression this PR introduced, and the diagnosis was exact.

`main` delivered the partial read — its `readStyles` swallowed failures in a `catch` and returned whatever it got:

```ts
} catch {
  // Style file not found, continue without this style
}
...
return styleContents.length > 0 ? styleContents : null
```

I over-corrected. `complete` guards exactly ONE thing: turning an unknown answer into a definitive `[]` that wipes live CSS. It is not a reason to throw away content that WAS read.

```ts
styles = external.complete || merged.length > 0 ? merged : null
```

```
complete                   -> merged, may be [], may clear   (new in this PR)
incomplete, merged non-[]  -> merged                         (= main)
incomplete, merged == []   -> null, key omitted              (= main)
```

Verified rather than argued: I transcribed both revisions' formulas and swept all 105 combinations of inline shape against read outcomes across 0-2 URLs. 21 differ from `main`, and all 21 are the intended "complete read, empty result -> `styles: []`". Zero other differences. The five cases in your scenario specifically:

```
2 externals, 1 fails, 1 has content    main=["a{}"]  fixed=["a{}"]   MATCH   (was null)
2 externals, 1 fails, 1 reads empty     main=null     fixed=null      MATCH
1 external, fails, inline empty         main=null     fixed=null      MATCH
1 external, fails, inline HAS content   main=["i{}"]  fixed=["i{}"]   MATCH   (was null)
1 external, fails, nothing else         main=null     fixed=null      MATCH
```

Three red-checked tests added. The first failed with `expected '// HMR update for: …' to contain 'PS457_PARTIAL_MARKER'`, the module carrying no `styles` key at all. Suite 410 -> 413; the clearing tests and all four "keeps the styles" tests still pass.

## On the retry recommendation

I did not implement retry or change the pending-update lifecycle, because the narrow fix removes the thing that needed retrying — the healthy stylesheet is now delivered on the first attempt. Walking the remaining cases:

- Only stylesheet transiently missing (atomic-write window): key omitted, old CSS survives, and the rename fires a second watcher event that re-queues a pending update.
- Only stylesheet permanently missing: nothing was read, so there is nothing to retry. Keeping the old CSS is this PR's contract for an unknown answer.

One limitation I am naming rather than fixing: two externals where one is permanently missing and the other is edited to empty leaves the key omitted and the old CSS in place. `main` did exactly this, and clearing there would be wrong while one sheet is unknown.
